### PR TITLE
Feature.robot

### PR DESCRIPTION
**Pull request by iakovgan**
This feature provides syntax hilighting of Robot Framework DSL. http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html

Comments, review and feedback is welcome. This is my first lexer. 

Thank you for the tool. 
Regards,
Iakov

### CODE REVIEW

File: lib/rouge/lexers/robot.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        rule %r([|]), Punctuation
+        rule /".+"/, Str::Double
+        rule /'.+'/, Str::Single
+        rule /".+"/, Str::Double
+        rule  /\b[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\b/, Str # ipv4
+#        rule /\W-?(?:0|[1-9]\d*)\.\d+(?:e[+-]\d+)?\W/i, Num::Float 
+#        rule /\W-?(?:0|[1-9]\d*)(?:e[+-]\d+)?\W/i, Num::Integer    
+      end
+
+      # Token types are set for visual perfofmance in gitlab webui
```
Comment (jneen) [inline lib/rouge/lexers/robot.rb]: *performance

File: lib/rouge/lexers/robot.rb [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        rule /\@\{.*?\}/, Name::Variable # list
+        rule /\&\{.*?\}/, Name::Variable # dict
+        rule %r([\r]+), Text
+        rule %r([|]), Punctuation
+        rule /".+"/, Str::Double
+        rule /'.+'/, Str::Single
+        rule /".+"/, Str::Double
+        rule  /\b[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\b/, Str # ipv4
+#        rule /\W-?(?:0|[1-9]\d*)\.\d+(?:e[+-]\d+)?\W/i, Num::Float 
+#        rule /\W-?(?:0|[1-9]\d*)(?:e[+-]\d+)?\W/i, Num::Integer
```
Comment (jneen) [inline lib/rouge/lexers/robot.rb]: Please don't check in commented code - do you mean to have these uncommented?

File: lib/rouge/lexers/robot.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      KEYWORD = 'Documentation|Tags|Arguments|Return|Teardown|Timeout'
+      rest_of_line = /.*?(?=[#\n])/
+
+
+      state :basic do
+#        #In robot framework tabs are very important, 
+#        #but highliting with a '→' sign can be an isssue for copypaste
+#        rule /\t/ do |m|
+#         #token Punctuation, '→' + m[0] 
+#        end
```
Comment (jneen) [inline lib/rouge/lexers/robot.rb]: Many preprocessors (such as redcarpet) do not preserve tabs in the output, but rather replace them with 4 spaces. Is it possible to highlight with that uncertainty in mind?

Comment (jneen) [inline lib/rouge/lexers/robot.rb]: (Also, it's a strong guarantee that rouge lexers output the same text that they are given, if you sum up all the tokens.)

Comment (iakovgan) [inline lib/rouge/lexers/robot.rb]: Thank you for the review. 

Small question : can we hilight empty spaces? Like background color? This would help to identify tabs and spaces?

**Comment by iakovgan** [comment]
@jneen , Thank you for the review. I've removed all unused code.

**Comment by michele-carignani** [comment]
Hi,

any updates on this PR? Is there a way to help moving this forward?

Thanks

**Comment by iakovgan** [comment]
To my eye this could be merged to trunk but it is up to manintainers to do.

Anyway with my branch I successfully patched my gitlab setup and it works of a year now. (Have to repatch on every major upgrade though).

**Comment by GLMeece** [comment]
Our team (using Gitlab) is starting to embrace Robot Framework. For a "replacement" of Pygments, this is a bit of a sore spot. 👊

**Comment by falxcerebri** [comment]
How's progress with this PR?

**Comment by pyrmont** [comment]
@falxcerebri I'm sorry you've been waiting so long :( I joined the maintainer's group a couple of months back and had been working my way through the outstanding PRs in reverse chronological order. Unfortunately, that means that the older PRs are having to wait even longer.

That said, while I follow the above as a general rule, I move PRs to the front of the queue when there's some kind of activity (like this!). I've added this one now and will be getting to it ASAP!

**Comment by pyrmont** [comment]
Sorry not to have made more rapid progress on this. I've forced pushed a rebased version that makes development a bit easier (once you've installed the gems, you can run tests with `rake`).

@iakovgan said it was his first lexer and unfortunately (although very understandably!) it looks to me like it requires a fair amount of rewriting to be more consistent with the way Rouge's regular-expression-based lexers usually work.

The difficulty (for me) is that the Robot framework is a DSL which means it's taking a bit of time for me to wrap my head around it. I'll try to move it forward as quickly as I can but progress will probably be a bit slower than usual :(

**Comment by pyrmont** [comment]
Just wanted to confirm I haven't forgotten about this: I'm afraid my progress has been slow but am still moving forward.

**Comment by falxcerebri** [comment]
That's great to hear! Fingers crossed!

**Comment by pyrmont** [comment]
It took a while but I've pushed a new version of the lexer.

I'm sure there are edge cases where it doesn't lex things correctly but it generally seems to work as expected. I know that it doesn't tokenise tests correctly where a template is set in the Settings section but that looks like a pain to handle (not least because those templates can be _unset_ per test) so I've left that for some unspecified point in the future.

Let me know what you think, @falxcerebri.
